extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,615 @@
+        this.handleFluidMessage_(payload);
+        break;
+      case SERVICE.EXPAND_REQUEST:
+        this.handleExpandRequest_(payload);
+        break;
+      case SERVICE.REGISTER_DONE:
+        this.handleRegisterDone_(payload);
+        break;
+      case SERVICE.COLLAPSE_REQUEST:
+        this.handleCollapseRequest_();
+        break;
+      default:
+        break;
+    }
+  }
+
+  /**
+   * @param {!JsonObject} payload
+   * @private
+   */
+  handleRegisterDone_(payload) {
+    this.initialHeight_ = payload['initialHeight'];
+    this.initialWidth_ = payload['initialWidth'];
+  }
+
+  /**
+   * @param {!JsonObject} payload
+   * @private
+   */
+  handleExpandRequest_(payload) {
+    let expandHeight = payload['expand_b'] + payload['expand_t'];
+    let expandWidth = payload['expand_r'] + payload['expand_l'];
+    // We assume that a fair amount of usage of this API will try to pass the final
+    // size of the iframe that is desired, instead of the correct usage which is
+    // specify how much to expand by. We try to protect against this, by detecting
+    // if the requested expansion size matches the size of the amp-ad element, and
+    // if so we assume that the desired effect was to resize to fill the element.
+    if (expandWidth != this.initialWidth_ ||
+        expandHeight != this.initialHeight_) {
+      expandWidth += Number(this.iframe_.width);
+      expandHeight += Number(this.iframe_.height);
+    }
+    this.handleSizeChange(expandHeight,
+        expandWidth,
+        SERVICE.EXPAND_RESPONSE);
+  }
+
+  /**
+   * @private
+   */
+  handleCollapseRequest_() {
+    this.handleSizeChange(this.initialSize_.height,
+        this.initialSize_.width,
+        SERVICE.COLLAPSE_RESPONSE,
+        /** isCollapse */ true);
+  }
+
+  /**
+   * @param {number} height
+   * @param {number} width
+   */
+  resizeIframe(height, width) {
+    if (this.iframe_) {
+      setStyles(this.iframe_, {
+        'height': height + 'px',
+        'width': width + 'px',
+      });
+    }
+  }
+
+  /**
+   * Resizes the safeframe, and potentially the containing amp-ad element.
+   * Then sends a response message to the Safeframe creative.
+   * @param {number} height In pixels.
+   * @param {number} width In pixels.
+   * @param {string} message
+   * @param {boolean=} optIsCollapse Whether this is a collapse attempt.
+   */
+  handleSizeChange(height, width, message, optIsCollapse) {
+    // If the new size is fully contained within the bounds of the amp-ad,
+    // we can resize immediately as there will be no reflow.
+    if (!optIsCollapse &&
+        width <= this.creativeSize_.width &&
+        height <= this.creativeSize_.height) {
+      this.resizeIframe(height, width);
+      this.sendResizeResponse(/** SUCCESS */ true, message);
+    } else {
+      this.resizeAmpAdAndSafeframe(
+          height, width, message, optIsCollapse);
+    }
+  }
+
+  /**
+   * @param {boolean} success
+   * @param {string} message
```

Comment:
can you rename message to messageType or similar?  I thought it was the actual content of the message

Reply:
done